purenix/src/Nix/Print.hs

```diff
@@ -82,74 +90,116 @@ style a _ Single = a
 style _ b Multi = b
 
 exprStyle :: ExprF Style -> Style
+exprStyle (Attrs _ [] []) = Single
+exprStyle (Attrs [] [(sty, _)] []) = sty
+exprStyle (Attrs [] [] [(_, sty)]) = sty
 exprStyle Attrs {} = Multi
 exprStyle Let {} = Multi
 exprStyle v = bool Single Multi $ elem Multi v
 
-exprPrec :: ExprF a -> Int
-exprPrec Var {} = 10
-exprPrec Num {} = 10
-exprPrec Path {} = 10
-exprPrec String {} = 10
-exprPrec Attrs {} = 10
-exprPrec List {} = 10
-exprPrec Sel {} = 9
-exprPrec App {} = 8
+newtype Precedence = Precedence Int deriving newtype (Num, Eq, Ord)
+
+data Associativity = AssocLeft | AssocRight | AssocNone
+  deriving (Eq, Show)
+
+exprAssoc :: ExprF a -> Associativity
+exprAssoc Sel {} = AssocLeft
+exprAssoc App {} = AssocLeft
+exprAssoc (Bin op _ _) = opAssoc op
+  where
+    opAssoc Update = AssocRight
```

Comment:
Huh, I'm surprised this associates to the right.

I was trying to figure out a way to test that this actually associates to the right in the Nix repl, but I couldn't come up with anything.  

I was thinking that because of the way the `//` operator works, any expression that is just a chain of `//` will always evaluate to the same thing, regardless of what associativity it has.  I guess it is a monoid.

Reply:
Yeah, I guess it's like `mappend` for maps, which is obviously associative?

Also, in the example generated code I posted you can see an expression consisting of only `&&`s in which one && term is parenthesized because it occurs in the right in the AST. I was thinking it might be worth adding an `Associative :: Associativity` to avoid some parens in these cases, that could also apply here.

Reply:
> I was thinking it might be worth adding an Associative :: Associativity to avoid some parens in these cases, that could also apply here.

Oh, that would be nice!

Reply:
Done in 8e7c93a248344a685d12a2a234f8421875514919